packages/gatsby/src/commands/develop.ts

```diff
@@ -1,683 +1,162 @@
-      }
-      process.exit(0)
-    }
+  const developProcessPath = resolveCwd.silent(
+    `gatsby/dist/commands/develop-process`
   )
+  // Run the actual develop server on a random port, and the proxy on the program port
+  // which users will access
+  const proxyPort = program.port
+  const developPort = await getRandomPort()
+
+  const script = createControllableScript(report.stripIndent`
+    const cmd = require("${developProcessPath}");
+    const args = ${JSON.stringify({
+      ...program,
+      port: developPort,
+      proxyPort,
+    })};
+    cmd(args);
+  `)
+
+  const proxy = startDevelopProxy({
+    proxyPort: proxyPort,
+    targetPort: developPort,
+    programPath: program.directory,
+  })
 
-  if (process.env.GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES) {
-    report.panic(
-      `The flag ${chalk.yellow(
-        `GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES`
-      )} is not available with ${chalk.cyan(
-        `gatsby develop`
-      )}, please retry using ${chalk.cyan(`gatsby build`)}`
-    )
-  }
-  initTracer(program.openTracingConfigFile)
-  report.pendingActivity({ id: `webpack-develop` })
-  telemetry.trackCli(`DEVELOP_START`)
-  telemetry.startBackgroundUpdate()
-
-  const port =
-    typeof program.port === `string` ? parseInt(program.port, 10) : program.port
-
-  // In order to enable custom ssl, --cert-file --key-file and -https flags must all be
-  // used together
-  if ((program[`cert-file`] || program[`key-file`]) && !program.https) {
-    report.panic(
-      `for custom ssl --https, --cert-file, and --key-file must be used together`
-    )
-  }
-
-  try {
-    program.port = await detectPortInUseAndPrompt(port)
-  } catch (e) {
-    if (e.message === `USER_REJECTED`) {
-      process.exit(0)
-    }
-
-    throw e
-  }
-
-  // Check if https is enabled, then create or get SSL cert.
-  // Certs are named 'devcert' and issued to the host.
-  if (program.https) {
-    const sslHost =
-      program.host === `0.0.0.0` || program.host === `::`
-        ? `localhost`
-        : program.host
-
-    if (REGEX_IP.test(sslHost)) {
-      report.panic(
-        `You're trying to generate a ssl certificate for an IP (${sslHost}). Please use a hostname instead.`
-      )
-    }
-
-    program.ssl = await getSslCert({
-      name: sslHost,
-      caFile: program[`ca-file`],
-      certFile: program[`cert-file`],
-      keyFile: program[`key-file`],
-      directory: program.directory,
-    })
-  }
-
-  // Start bootstrap process.
-  const { graphqlRunner } = await bootstrap(program)
-
-  // Start the createPages hot reloader.
-  bootstrapPageHotReloader(graphqlRunner)
-
-  // Start the schema hot reloader.
-  bootstrapSchemaHotReloader()
-
-  await queryUtil.initialProcessQueries()
+  const wsServerPort = await getRandomPort()
 
-  require(`../redux/actions`).boundActionCreators.setProgramStatus(
-    `BOOTSTRAP_QUERY_RUNNING_FINISHED`
+  const success = await createServiceLock(
+    program.directory,
+    `ws`,
```

Comment:
[nit] This isn't a websocket service, it is a devserver status service, I think the name should be semantic, not implementation based.